### PR TITLE
Silent publishing

### PR DESCRIPTION
**Pull request by tokejepsen**
This PR gives a built-in option for end-users to publish without any GUI.

oct 26 2017 1-24 pm
oct 26 2017 1-25 pm

### CODE REVIEW

**Comment by mottosso** [comment]
I think for this to work, there will have to be some indicator progress. Otherwise you'd have no way of knowing whether it froze or just takes a really long time.

It'd be simple enough to do, we can convert `util.publish` into an iterator, and have it emit `result` dictionaries per plug-in. Then your splash gui could simply print the currently processed plug-in and perhaps overall status (errored, or all-clear).

**Comment by tokejepsen** [comment]
With the publish iterator on its way, let's talk about how to show the progress.

Initially I was thinking of a simple label of the plugin plus success/failed icon:

publish_progress

Ideally a progress bar would be better, but getting a total count of the plugins won't be possible without collecting first, unless you would count collection as a 1/4 of the progress.

**Comment by mottosso** [comment]
> Ideally a progress bar would be better, but getting a total count of the plugins won't be possible without collecting first, unless you would count collection as a 1/4 of the progress.

That's a good point. 

Maybe it'd be ok to yield different results at different times? For example..

```python
iterator = pyblish.util.publish_iter()
plugins, context = next(iterator)  # Before collection
plugins, context = next(iterator)  # After collection, but before main processing
for Plugin, context in iterator:
  ...
```

Not sure how convenient that is..

**Comment by tokejepsen** [comment]
> Maybe it'd be ok to yield different results at different times? For example..

Not entirely sure what you mean here.

We could also have two progress bars; one for collection and one for the main processing?
Maybe that is what you meant...

**Comment by mottosso** [comment]
> Not entirely sure what you mean here.

I mean yielding at multiple times in the function. It doesn't just have to come from the loop.

```python
def publish():
  context = api.Context()
  yield context
  for collector in [SomeCollector]:
    api.process(collector)
  yield context

  for Plugin, instance in api.discover():
    yield Plugin, instance
```

Now when you call `next(iterator)` you'll get the first yield. When you call it again, you'll get the second yield. The third time onwards, you'll get the results from the loop.

**Comment by tokejepsen** [comment]
Alright let's walk through an example as I see it working.

I have ten plugins; three collectors, three validator, three extractors and one integrator.

- Initially all plugins are counted as the total, but 0 is processed:
    - ```0/10 = 0```
- The collection is run, so 3 plugins are getting processed:
    - ```1/10 = 0.1```
    - ```2/10 = 0.2```
    - ```3/10 = 0.3```
- The collection is done and a validator and extractor gets disregarded because of missing instances:
    - ```3/8 = 0.375```
- The remaining plugins get processed:
    - ```4/8 = 0.5```
    - ```5/8 = 0.625```
    - ```6/8 = 0.75```
    - ```7/8 = 0.875```
    - ```8/8 = 1```

Depending on the amount of plugins disregarded, the jump between the collection increment and the rest of the increments can between quite large.
Also need to note that we are not considering the amount of instances a plugin needs to be process, but I can't figure out how to incorporate that.

**Comment by tokejepsen** [comment]
Actually I might have a solution for considering the instances.

Until the collection has been processed the percentage calculation would be:

```
(plugins processed) / (total plugins to process)
```

Once we have processed the collection, we'll have some instances. Continue from the above example the collection produced 4 instances that will be processed by the remaining plugins with the following calculation:

```
((instances processed / amount of instances) + plugins processed)) / (total plugins to process)
```
- ```((1/4) + 3)/8 = 0,40625```
- ```((2/4) + 3)/8 = 0,4375```
- ```((2/4) + 3)/8 = 0,46875```
- ```((2/4) + 3)/8 = 0,5```

etc...

Will all this fairly complicated code, we are probably best off having the percentage calculation yielded from the publish iterator so the returns of the yields would be ```(percentage, result, context)```

**Comment by mottosso** [comment]
> Will all this fairly complicated code, we are probably best off having the percentage calculation yielded from the publish iterator so the returns of the yields would be (percentage, result, context)

How about we add `progress` to the result dictionary? It's already got a `duration` entry, for how long it took a particular plug-in to run, so I think some more of that kind of metadata would fit.

**Comment by tokejepsen** [comment]
> How about we add progress to the result dictionary?

Sure, that sounds good to me. Would it be alright to add that in the ```publish.util``` module only?

**Comment by mottosso** [comment]
I guess for starters, add it like you did, and see if that's enough to build a reasonable GUI. If we add to `result` we'll need to do that across the board, for GUIs as well.

**Comment by tokejepsen** [comment]
Added a progress bar to work with the new ```pyblish.util.publish_iter```

oct 31 2017 3-49 pm

**Comment by tokejepsen** [comment]
What do you think of this now?
Works with https://github.com/pyblish/pyblish-base/pull/326

File: pyblish_nuke/lib.py [hunk 1]
```diff
@@ -144,10 +144,88 @@ def add_to_filemenu():
...
+
+        messagebox.exec_()
+
+    callback = "published", on_published
+    pyblish.api.register_callback(*callback)
+
+    def publish_iter():
+
+        for result in pyblish.util.publish_iter():
+            splash.bar.setValue(result["percentage"] * 100)
```
Comment (mottosso) [inline pyblish_nuke/lib.py]: How come your multiplying the percentage by 100? Would it already be from 1-100? Also, I didn't catch this, but percentage is not a good name for this. A percentage of what? Let's call it `progress` and have it go from 0-1.

Comment (tokejepsen) [reply pyblish_nuke/lib.py]: Yeah, I guess ```percentage``` indicates a 0-100 range.

I'll change the name to ```progress```, as the calculation is already in the 0-1 range.

Comment (tokejepsen) [reply pyblish_nuke/lib.py]: Done :)

**Comment by tokejepsen** [comment]
Is this mergable now?

When it is I'll focus on some tests in pyblish-base to cement the essential parts of the functionality.

**Comment by mottosso** [comment]
We can't merge this, as the feature doesn't exist in base yet.

**Comment by tokejepsen** [comment]
> We can't merge this, as the feature doesn't exist in base yet.

Sorry, didn't mean for it to be merged, just whether it was looking complete to you.

File: pyblish_nuke/lib.py [hunk 2]
```diff
@@ -144,10 +144,88 @@ def add_to_filemenu():
...
+    splash.show()
+
+    def on_published(context):
+        try:
+            splash.close()
+        except RuntimeError:
+            # Splash already closed
+            pass
+
+        pyblish.api.deregister_callback(*callback)
```
Comment (mottosso) [inline pyblish_nuke/lib.py]: Better deregister this *before* closing the window, in case a different exception is thrown.

**Comment by mottosso** [comment]
Other than that nitpick, sure, looks good!

**Pull request merged**